exaile.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH EXAILE 1 "October 16, 2006"
.SH NAME
exaile \- an AmaroK-like media player
.SH SYNOPSIS
.B exaile
.RI [options]
.SH DESCRIPTION
.B Exaile
is a media player aiming to be similar to KDE's AmaroK, but for
GTK+. It incorporates many of the cool things from AmaroK (and other media
players) like automatic fetching of album art, handling of large libraries,
lyrics fetching, artist/album information via the wikipedia, last.fm
support, optional iPod support (assuming you have python-gpod installed).
 
.SH OPTIONS
.TP
.B \-h, \-\-help
Show summary of options and exit
.TP
.B \-d DIR, \-\-duplicates=DIR
Finds and deletes all duplicate tracks (based on their md5 sum)
.TP
.B \-n, \-\-next
Play the next track
.TP
.B \-p, \-\-prev
Play the previous track
.TP
.B \-s, \-\-stop
Stop playback
.TP
.B \-a, \-\-play
Play
.TP
.B \-t, \-\-play\-pause
Play/Pause Playback
.TP
.B \-\-play\-cd
Play A CD
.TP
.B \-q, \-\-query
Query player
.TP
.B \-\-gui\-query
Show a popup of the currently playing track
.TP
.B \-\-get\-title
Print the title of current track
.TP
.B \-\-get\-album
Print the album of current track
.TP
.B \-\-get\-artist
Print the artist of current track
.TP
.B \-\-get\-length
Print the length of current track
.TP
.B \-\-current\-position
Print the position inside the current track as a percentage
.TP
.B \-i VOL, \-\-increase_vol=VOL
Increases the volume by VOL
.TP
.B \-l VOL, \-\-decrease_vol=VOL
Decreases the volume by VOL
.TP
.B \-\-stream=URL
Stream URL
.TP
.B \-\-new
Start new instance
.TP
.B \-\-settings=SETTING
Settings Directory

.SH SEE ALSO
More information is available on the
.B exaile
web page:
.IR http://www.exaile.org
.SH AUTHOR
.B Exaile
was written by Adam Olsen <arolsen@gmail.com>
.P
This manual page was written by Fran\[,c]ois F\['e]votte <francois.fevotte@ensta.org>


changelog
Exaile 0.2.10 (Friday June 22, 2007)

  * #1 (enhancement) Graphic Equalizer
  * #192 (bug) Doubled slash in Files tab
  * #267 (bug) metatag found foundcallback and i18n
  * #286 (enhancement) Translation: pa (Punjabi)
  * #288 (bug) Mini mode: scrolling over dropdown stops playback
  * #290 (bug) Library will not recognise WMA!
  * #304 (bug) Xgettext complains about unnamed arguments
  * #316 (bug) Makefile does not install locales
  * #319 (bug) Can not play music when adding a qeue immediately after starting exaile
  * #320 (task) CLOSED Translation: en_GB (British English)
  * #323 (bug) Mini mode and main window visible at the same time
  * #324 (bug) Exaile should not appears on all virtal desktops
  * #330 (bug) ImportError: libgtkembedmoz.so:
  * #341 (bug) mmkeys.so, ELF shared library is installed to /usr/share instead of /usr/lib.
  * #342 (task) Better volume handling
  * #349 (enhancement) Turkish translation
  * #350 (bug) information -> artist = Crash
  * #351 (enhancement) Compile to .pyc to speed up loading.
  * #354 (bug) Wont open files with Brazilian Portuguese Characters (PT_BR)
  * #358 (bug) Makefile is broken
  * #361 (bug) Change of $PREFIX breaks some stuff
  * #363 (bug) lib/* not installed
  * #365 (bug) not supporting new libatk
  * #369 (enhancement) Added German translations to exaile.desktop
  * #371 (enhancement) Sonata-like progress bar
  * #373 (bug) Available plugin bug
  * #376 (bug) OSD pops out unwanted
  * #378 (bug) Translation doesn't work
  * #379 (bug) Libnotify over tray icon
  * #381 (bug) Just little bugs when exaile start
  * #382 (enhancement) Set a window-role/title to desktop-cover window to enable full X11 window management
  * #384 (bug) visualization bug
  * #393 (enhancement) Italian Translation
  * #396 (bug) OSD shows on startup even though OSD is disabled
  * #398 (bug) Remove all installed files on uninstall
  * #399 (bug) Multimedia keys not working in Ubuntu Feisty
  * #404 (bug) Glitches in mini mode: progress slider & remaining time
  * #405 (bug) Exaile -a or --play doesn't work, instead it uses exaile --next
  * #406 (enhancement) Use PNG rating star image
  * #410 (bug) Command line doesn't work
  * #423 (bug) Tracks disappear from playlist when filename contains the: # pound hash sign
  * #426 (enhancement) CLOSED Translation: sv (Swedish)
  * #435 (bug) No tray icon and notification displaced
  * #440 (bug) Exaile won't start if last used drive (e.g., external drive) is missing
  * #447 (bug) Translations not found
  * #450 (bug) Packaging problem due to new generation of 'exaile' script
  * #451 (bug) Strings in the plugin manager window are not translated
  * #452 (bug) Firefox fix breaks debian packaging
  * #453 (enhancement) Translation: zh_TW (Traditional Chinese)
  * #460 (bug) Library Manager window's title
  * #461 (bug) "Cancel" and "Choose" in the import dialog do not work correctly
  * #462 (bug) The strings in the preferences TreeView in the Preferences Window are not translatable
  * #465 (bug) The strings in the "Edit Track(s)" windows are not translatable
  * #467 (enhancement) Added 'Show track in Collection' functionality
  * #471 (bug) Switch --version yields error
  * #472 (bug) error plugin configuration
  * #475 (bug) Tray icon menu not translated
  * #476 (enhancement) Exaile cannot start and play a CD
  * #477 (bug) exaile_xchat.pl should catch if there is no Artist for track
  * #479 (bug) Files tab  is broken is gstreamer-gnome-vfs is missing
  * #481 (enhancement) Translation: pt_BR (Brazilian Portuguese)
  * #490 (bug) Opening files from Nautilus or terminal makes Exaile open, but it does not load song
  * #495 (bug) "Open URL" and a URL like "http://shoutcast.com/something.pls" doesn't work
  * #499 (enhancement) Translation: es_AR (Argentine Spanish)
  * #502 (enhancement) zh_TW translation of Exaile
  * #503 (enhancement) Translation: pl (Polish)
  * #504 (enhancement) Translation: cs (Czech)
  * #513 (bug) Playback bugs in minimode
  * #517 (bug) Interface not translated
  * #518 (bug) SPAT: broken image displayed at startup
  * #519 (enhancement) Search for song title
  * #526 (bug) Volume control on tray icon/OSD popup broken
  * #527 (enhancement) Change name to make Exaile program menu option more meaningful
  * #529 (enhancement) Translation: nn (Nynorsk)
  * #530 (bug) exaile / exaile.py doens't work !
  * #531 (bug) Genre ID3 tags not picked up in WMA tracks
  * #533 (bug) Tray icon
  * #534 (enhancement) update to "did you know..." plugin
  * #537 (enhancement) feature request: hiding of the sidebar
  * #547 (enhancement) Changing the color and the size of the rating star
  * #550 (enhancement) Translation: hu_HU (Hungarian)
  * #551 (enhancement) Consider use of lngettext()
  * #553 (enhancement) Translation: ro (romanian)
  * #555 (bug) OSD - Changing volume doesn't update when Exaile doesn't have focus
  * #556 (enhancement) Patch to fix inconsistencies in the SPAT icons.
  * #557 (task) Translation: de (German)
  * #558 (bug) Error when closing Visualizations window when music is not playing.
  * #559 (bug) Various UTF-8 issues when rescanning collection.
  * #560 (enhancement) Translation: en_GB (British English)
  * #569 (task) Translation: he_IL (Hebrew)
  * #572 (enhancement) Change of queue icon on the right click menu.
  * #577 (enhancement) Simplifications for the createpot script
  * #582 (enhancement) he_IL translation
  * #584 (bug) Title and Artist field not updated correctly when listening a Shoutcast radio in a localized Exaile
  * #600 (enhancement) Alphabetic grouping: The ability to turn it off if you don't like it.
  * #615 (enhancement) Use gtkmozembed instead of gtkhtml2


exaile.py
#!/usr/bin/env python

# Copyright (C) 2006 Adam Olsen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 1, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

import sys

if sys.platform == 'linux2':
    # Set process name.  Only works on Linux >= 2.1.57.
    try:
        import dl
        libc = dl.open('/lib/libc.so.6')
        libc.call('prctl', 15, 'exaile\0', 0, 0, 0) # 15 is PR_SET_NAME
    except:
        pass

import gobject
gobject.threads_init()

# this stuff is done first so that only the modules required to connect to an
# already loaded exaile (if available) are loaded.  This helps with the speed
# of remote commands, like --next, --prev
import xl.dbusinterface
EXAILE_OPTIONS = xl.dbusinterface.get_options()
DBUS_EXIT = xl.dbusinterface.test(EXAILE_OPTIONS)

# find out if they are asking for help
HELP = False
for val in sys.argv:
    if val == '-h' or val == '--help': HELP = True

if '-h' in sys.argv: sys.argv.remove('-h')
if '--help' in sys.argv: sys.argv.remove('--help')
if '--version' in sys.argv:
    print "Exaile version: %s" % __version__
    sys.exit(0)

import os.path

import pygtk
pygtk.require('2.0') # Must be before 'import gtk'
import gtk

## Find out the location of exaile's working directory, and go there
basedir = os.path.dirname(os.path.realpath(__file__))
if not os.path.exists(os.path.join(basedir, "exaile.py")):
    if os.path.exists(os.path.join(os.getcwd(), "exaile.py")):
        basedir = os.getcwd()
sys.path.insert(0, basedir)
os.chdir(basedir)

# Heuristic to check whether Exaile is installed.
# Sets prefix accordingly, otherwise it is ".".
prefix = '.'
path_suffix = '%sshare%sexaile' % (os.sep, os.sep)
if basedir.endswith(path_suffix):
    prefix = basedir[:-len(path_suffix)]
    # Add ../../lib/exaile to path, otherwise Exaile won't be able to import any
    # of its modules.
    sys.path.append(os.path.join(prefix, 'lib', 'exaile'))

# set up gettext for translations
import gettext, locale
import gtk.glade
locale.setlocale(locale.LC_ALL, '')
gettext.textdomain('exaile')
gtk.glade.textdomain('exaile')
if prefix == '.': # if Exaile is not installed
    gettext.bindtextdomain('exaile', 'po')
    gtk.glade.bindtextdomain('exaile', 'po')
else:
    gtk.glade.bindtextdomain('exaile', os.path.join(prefix, 'share', 'locale'))

if sys.platform.startswith("win"):
    sys_var = "USERPROFILE"
else:
    sys_var = "HOME"
gtk.window_set_default_icon_from_file("images%sicon.png"% os.sep)
SETTINGS_DIR = "%s%s%s" % (os.getenv(sys_var), os.sep, ".exaile")
GCONF_DIR = "/apps/exaile"

from xl.gui import main as exailemain
from xl import xlmisc

def check_dirs():
    """
        Makes sure the required directories have been created
    """
    covers = os.path.join(SETTINGS_DIR, "covers")
    cache = os.path.join(SETTINGS_DIR, "cache")
    if not os.path.isdir(covers):
        os.mkdir(covers)

    if not os.path.isdir(cache):
        os.mkdir(cache)
    
def first_run(): 
    """
        Called if the music database or settings files are missing.

        Creates the settings directory, and, if necessary, creates the initial
        database file.

        Also scans the default import directory in case there's any music files
        there.
    """
    try:
        os.mkdir(SETTINGS_DIR)
    except:
        print "Could not create settings directory"

def main(): 
    """
        Everything dispatches from this main function
    """
    global SETTINGS_DIR
    p = EXAILE_OPTIONS

    if HELP:
        p.print_help()
        sys.exit(0)

    options, args = p.parse_args()
    if options.settings:
        SETTINGS_DIR = options.settings
    elif options.dups:
        xlmisc.log("Searching for duplicates in: %s" % options.dups)
        track.find_and_delete_dups(options.dups)
        sys.exit(0)

    running_checks = ('next', 'prev', 'stop', 'play', 'guiquery', 'get_title',
        'get_artist', 'get_album', 'get_length', 'current_position',
        'inc_vol', 'dec_vol', 'query')


    # check passed arguments for options that require exaile to currently be
    # running
    if not DBUS_EXIT:
        for check in running_checks:
            if getattr(options, check):
                print "No running Exaile instance found."
                sys.exit(1)

        fr = False
        if not os.path.isdir(SETTINGS_DIR):
            first_run()
            fr = True


        check_dirs()
        
        exaile = exailemain.ExaileWindow(SETTINGS_DIR, options, fr)
    else:
        sys.exit(0)

    gtk.main()

if __name__ == "__main__": 
    try:
        main()
    except SystemExit:
        raise
    except: # BaseException doesn't exist in python2.4
        import traceback
        traceback.print_exc()
        xlmisc.log_exception()
